### PR TITLE
#758 Mark submitted file errors

### PR DESCRIPTION
**Pull request by demariadaniel**
**Description of changes**

Clinical Service / Platform API / Platform UI updates to query Program Clinical Entity Data & Errors

PR's:
https://github.com/icgc-argo/platform-ui/pull/2284
https://github.com/icgc-argo/platform-api/pull/548
https://github.com/icgc-argo/argo-clinical/pull/762

Issues:
#539: https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/icgc-argo/platform-api/539

#758: https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/icgc-argo/argo-clinical/758

#540: https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/icgc-argo/platform-api/540

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [ ] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/clinical/clinical-api.ts [hunk 1]
```diff
@@ -61,6 +61,20 @@ class ClinicalController {
     res.send(zip.toBuffer());
   }
 
+  @HasProgramReadAccess((req: Request) => req.params.programId)
```
Comment (joneubank) [inline src/clinical/clinical-api.ts:64]: 👍

File: src/clinical/clinical-api.ts [hunk 2]
```diff
@@ -61,6 +61,20 @@ class ClinicalController {
...
+  async getProgramClinicalData(req: Request, res: Response) {
+    const programId = req.params.programId;
+    if (!programId) {
+      return ControllerUtils.badRequest(res, 'Invalid programId provided');
```
Comment (joneubank) [inline src/clinical/clinical-api.ts:68]: not so much invalid as "missing"

Comment (joneubank) [reply src/clinical/clinical-api.ts:68]: this is consistent with the parameter checking of the other methods. comment withdrawn.

File: src/clinical/clinical-api.ts [hunk 3]
```diff
@@ -61,6 +61,20 @@ class ClinicalController {
...
+    }
+
+    const data = await service.getClinicalData(programId);
+
+    res.status(200).contentType('application/json');
+
+    res.send(data);
```
Comment (joneubank) [inline src/clinical/clinical-api.ts:75]: `res.status(200).json(data);`

File: src/clinical/service-worker-thread/tasks.ts [hunk 4]
```diff
@@ -49,13 +49,13 @@ function extractDataFromDonors(donors: Donor[], schemasWithFields: any) {
...
     Object.values(ClinicalEntitySchemaNames).forEach(entity => {
-      let clincialInfoRecords;
+      let clinicalInfoRecords;
       if (entity === ClinicalEntitySchemaNames.REGISTRATION) {
-        clincialInfoRecords = getSampleRegistrationDataFromDonor(d);
+        clinicalInfoRecords = getSampleRegistrationDataFromDonor(d);
       } else {
-        clincialInfoRecords = getClinicalEntitiesFromDonorBySchemaName(d, entity);
+        clinicalInfoRecords = getClinicalEntitiesFromDonorBySchemaName(d, entity);
       }
```
Comment (joneubank) [inline src/clinical/service-worker-thread/tasks.ts:57]: thanks for typo fix... would probably write this like:
```
const clinicalInfoRecords = entity === ClinicalEntitySchemaNames.REGISTRATION
    ? getSampleRegistrationDataFromDonor(d)
    : getClinicalEntitiesFromDonorBySchemaName(d, entity);
```

Comment (demariadaniel) [reply src/clinical/service-worker-thread/tasks.ts]: Updated

File: src/resources/swagger.yaml [hunk 5]
```diff
@@ -771,7 +771,26 @@ paths:
         '500':
           $ref: '#/components/responses/ServerError'
         '200':
-          description: 'Stream of newline `\n` delmited JSON objects'
+          description: 'Stream of newline `\n` delimited JSON objects'
+
+  /clinical/program/:programId/clinical-data:
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/resources/swagger.yaml:775]: Good call, updated! I've also added the Clinical Error endpoint

File: src/clinical/service-worker-thread/tasks.ts [hunk 6]
```diff
@@ -79,10 +58,58 @@ function extractDataFromDonors(donors: Donor[], schemasWithFields: any) {
...
+
+      const relevantSchemaWithFields = schemasWithFields.find((s: any) => s.name === entityName);
+      if (!relevantSchemaWithFields) {
+        throw new Error(`Can't find schema ${entityName}, something is wrong here!`);
+      }
+
+      return {
+        entityName,
+        records,
+        entityFields: ['donor_id', ...relevantSchemaWithFields.fields],
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/clinical/service-worker-thread/tasks.ts]: Updated

File: src/clinical/donor-repo.ts [hunk 7]
```diff
@@ -152,6 +161,47 @@ export const donorDao: DonorRepository = {
     return F(mapped);
   },
 
+  async findByPaginatedProgramId(
+    programId: string,
+    query: ClinicalQuery,
+  ): Promise<DeepReadonly<Donor[]>> {
+    const { page, limit, sort, entityTypes, donorIds, submitterDonorIds, completionState } = query;
+
+    const projection = [
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/clinical/donor-repo.ts]: Updated

**Comment by blabadi** [comment]
If you can add a happy path integration test at least that would be great, it will help in avoiding regressions in the service

File: src/clinical/clinical-api.ts [hunk 8]
```diff
@@ -31,6 +31,31 @@ import { Donor } from './clinical-entities';
 import { omit } from 'lodash';
 import { DeepReadonly } from 'deep-freeze';
 
+export type ClinicalQuery = {
+  page: number;
+  limit: number;
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/clinical/clinical-api.ts:36]: Limit is likely 100 entries, similar to in the file-repo

File: src/clinical/clinical-service.ts [hunk 9]
```diff
@@ -163,12 +170,68 @@ export const getClinicalData = async (programId: string) => {
...
+
+  // worker-threads can't get dictionary instance so deal with it here and pass it to worker task
+  const allSchemasWithFields = await dictionaryManager.instance().getSchemasWithFields();
+
+  // async/await functions just hang in current library worker-thread setup, root cause is unknown
+  const donors = await donorDao.findByPaginatedProgramId(programId, query);
+
+  const taskToRun = WorkerTasks.ExtractEntityDataFromDonors;
+  const taskArgs = [donors, allSchemasWithFields];
+  const data = await runTaskInWorkerThread<ClinicalEntityData>(taskToRun, taskArgs);
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/clinical/clinical-service.ts:193]: Good question, I'm not sure. I'm following the pattern set in `getClinicalData`, which requests all Donors for a program. `getPaginatedClinicalData` at least has the limits set by the pagination, but could still have up to 100 results. 

Neither worker task is doing anything asynchronous so I'm not sure how necessary the worker threads are.

**Comment by demariadaniel** [comment]
> If you can at least add a couple happy path integration test for this new API, it would be great, it will help in avoiding regressions in the service

Will do! Once I resolve why 2 of our integration tests are failing....

File: src/clinical/clinical-api.ts [hunk 10]
```diff
@@ -31,6 +31,31 @@ import { Donor } from './clinical-entities';
 import { omit } from 'lodash';
 import { DeepReadonly } from 'deep-freeze';
 
+export type ClinicalQuery = {
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/clinical/clinical-api.ts:34]: Updated

File: src/clinical/clinical-service.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/clinical/clinical-service.ts]: I've tested this with 100 entries and don't think speed will be an issue. I've refactored the service to call this function without using the service worker. Turns out the changes to `common-model/functions` were what was breaking the integration tests. So this is in a much better place now.

**Comment by demariadaniel** [comment]
> If you can at least add a couple happy path integration test for this new API, it would be great, it will help in avoiding regressions in the service

Added basic identity integration tests @blabadi!

**Review by blabadi — APPROVED** [review]

**Pull request merged**